### ISSUE TITLE
get_cloest_lane_index becomes bottleneck when the map is large

### ISSUE THREAD

**Issue by pengzhenghao** [decisionforce/pgdrive#205]
```python
import time

from pgdrive import PGDriveEnv

if __name__ == '__main__':
    env = PGDriveEnv(dict(environment_num=10, map=7, traffic_density=0.2))
    obs = env.reset()
    start = time.time()
    action = [0.0, 0.1]
    for s in range(10000):
        o, r, d, i = env.step(action)
        if d:
            env.reset()
        if (s + 1) % 100 == 0:
            print(
                "Finish {}/10000 simulation steps. Time elapse: {:.4f}. Average FPS: {:.4f}".format(
                    s + 1,
                    time.time() - start, (s + 1) / (time.time() - start)
                )
            )
    print(f"Total Time Elapse: {time.time() - start}")
```

Suppose we use large map to profile, the FPS becomes half of 3-block case.

The most time-consuming operation is the `get_cloest_lane_index`.

Here is the profile result:

![pgdrive7](https://user-images.githubusercontent.com/22206995/103296703-95ae1980-4a31-11eb-9141-36582567b5a2.png)

I think we should, at the end, optimize this issue.

**Comment by lqy0057** [comment]
I have an idea about it by using rayCastTest() to query the lane index. I will try this when we need to optimize it

**Comment by pengzhenghao** [comment]
You mean a laser ray that shots to the ground? That's excellent idea!

**Comment by lqy0057** [comment]
Yes, after doing this the efficiency will not be affected by the number of lanes and the area of map. Then the bottleneck will be the traffic system,  so we have to rewrite a new traffic system for PGDrive by using BulletVehicle....

**Comment by pengzhenghao** [comment]
<img width="2096" alt="Screen Shot 2021-01-05 at 17 11 23" src="https://user-images.githubusercontent.com/22206995/103627973-46a83d00-4f79-11eb-95ce-e6a5d3f22606.png">

This would be a critical issue in future. 

You can see that the 3, 5, 7 blocks environments take 75k, 100k, 122k seconds to train, respectively.

This will limit future extension and scalability of PGDrive.

**Comment by lqy0057** [comment]
This will be solved soon by adding virtual body for lane, and use RayCaset to query the lane information

**Comment by pengzhenghao** [comment]
> This will be solved soon by adding virtual body for lane, and use RayCaset to query the lane information

No..

I turn my point now. I think a ray caset to the road is not a good practice. This will cause chaos when you are in a 3D world.

My suggestion is to maintain a 3D world-coordinate bounding box system. Each movable element should register its location to the bounding box manager. So the bounding box manager can maintain a mapping between element id to the bounding box id. Only the elements in the same bounding box can interactive with each other.

(BTW, do we have a movable element id system? It seems that we just maintain everything in seperated list in the scene manager. I think a general way is to maintain the "element list", instead of treating different element differently, like some are in traffic_vehicles list and some are in other list.)

**Comment by pengzhenghao** [comment]
@lqy0057 Does your latest PR can also consider what I have say here? We need to think about how to scale up in future.

**Comment by lqy0057** [comment]
resolved in #237

**Comment by lqy0057** [comment]
#237 Mergerd, close
